Generator of Tendermint types for unit, integration, and model-based testing

andrey-kuprianov opened the pull request:
Here we introduce both the code-level helpers for generating tendermint object hierarchies in tests, as well as the command-line generator of those hierarchies.

For the high-level explanation please see #393.

This is still a WIP, but as it is already usable it probably makes sense to merge now, and then extend in separate PRs/issues.

Will be grateful for your reviews!

* [X] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [X] Updated all code comments where relevant
* [X] Wrote tests
* [ ] Updated CHANGES.md

shonfeder commented on tendermint-testgen/src/helpers.rs:
```diff
@@ -0,0 +1,83 @@
+    }
+}
+
+pub fn choose_from<T: Clone>(cli: &Option<T>, input: &Option<T>) -> Option<T> {
```
iiuc, this just [Option::or](https://doc.rust-lang.org/std/option/enum.Option.html#method.or) with a clone at the end? So I wonder if you might just use `foo.or(bar).map(|x| x.clone)`, or at least implement this in terms of that? Personall, id be in favor of the first option, since it's nice for readers to be able to think in terms of standard library functions when possible, rather than having to lookup idiosyncratic, shallow wrappers around them.

If the latter, then this function should perhaps be renamed to `option_or_clone` or something, so it's clear how this is related to `Option::or`, with the addition of a clone.

andrey-kuprianov replied:
very nice! I was simply unaware of that function; replaced my homegrown one with this everywhere...

shonfeder commented on tendermint-testgen/src/helpers.rs:
```diff
@@ -0,0 +1,83 @@
+    }
+}
+
+pub fn choose_or<T>(input: Option<T>, default: T) -> T {
```
I think this is identical to https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or, and that should be used instead.

andrey-kuprianov replied:
thanks! same as above: was not aware of this. replaced everywhere

shonfeder commented on tendermint-testgen/src/helpers.rs:
```diff
@@ -0,0 +1,83 @@
+use signatory_dalek::Ed25519Verifier;
+use signatory::signature::Verifier;
+
+#[macro_export]
```
Some documentation for tihs would be helpful.

andrey-kuprianov replied:
done

shonfeder replied:
Thanks!

brapse commented on tendermint-testgen/src/commit.rs:
```diff
@@ -0,0 +1,159 @@
+
+    // get a mutable reference to the vote of the given validator
+    // this function will panic if the votes or the validator vote is not present
+    pub fn vote_of(&mut self, val: &str) -> &mut Vote {
```
perhaps `vote_of_validator` is a better naming.

andrey-kuprianov replied:
thanks for the name suggestion, applied

brapse commented:
Went through this at a high level so excuse the lack of detail but in general I would say this is pretty great.

Besides the thoughtful refinements mentioned by @shonfeder I would look forward to this landing sooner rather than later. As a validation and natural next step I would look to validate the capabilities and abstractions by making progress on #330. Perhaps adding a generator for light blocks might help there.

andrey-kuprianov commented on tendermint-testgen/bin/tendermint-testgen.rs:
```diff
@@ -0,0 +1,129 @@
+    producer.encode()
+}
+
+fn run_command<Opts: Generator<T> + Options, T: serde::Serialize>(cli: Opts, read_stdin: bool) {
```
well, I also considered using `Abscissa` first, but then I saw that `light-client/examples/light-client.rs` uses `Gumdrop`, and I liked it as more easy to use, at least for me, at my present state of knowledge. It doesn't introduce almost any additional overhead, and please also consider that it allows me to employ the same structs both for Rust API, and for CLI command options, which I think is very neat.

Let's leave it as is for now, and we'll have a chance to reconsider later.

romac replied:
Note: `abscissa` uses `gumdrop` as well, but comes with more bells and whistle which I agree might not be needed for this specific project. As you say, we can still consider moving to `abscissa` a later point if the CLI grows or undergoes substantial changes.

shonfeder replied:
I was assuming Abscissa would handle stuff like reporting supported arguments and error return codes. But agree there's no reason you have to do it here and now -- or at all :)

andrey-kuprianov commented on tendermint-testgen/bin/tendermint-testgen.rs:
```diff
@@ -0,0 +1,129 @@
+        }
+        Some(Command::Validator(cli)) => run_command(cli, opts.stdin),
+        Some(Command::Header(cli)) => run_command(cli, opts.stdin),
+        Some(Command::Vote(cli)) => eprintln!("{:?}", cli),
```
thanks for spotting this! it is actually implemented, only this last step was missing...

andrey-kuprianov commented on tendermint-testgen/src/commit.rs:
```diff
@@ -0,0 +1,159 @@
+    set_option!(votes, Vec<Vote>);
+    set_option!(round, u64);
+
+    // generate commit votes from all validators in the header
```
Nice suggestion! Added these and other doc comments.

andrey-kuprianov commented on tendermint-testgen/src/commit.rs:
```diff
@@ -0,0 +1,159 @@
+
+    // get a mutable reference to the vote at the given index
+    // this function will panic if the votes or the vote at index is not present
+    pub fn vote_at(&mut self, index: usize) -> &mut Vote {
```
thanks for the name suggestion, applied also

romac commented on tendermint-testgen/src/header.rs:
```diff
@@ -84,15 +84,15 @@ impl Generator<block::Header> for Header {
-            height: block::Height(choose_or(self.height, 1)),
-            time: choose_or(self.time, Time::now()),
+            height: block::Height(self.height.unwrap_or(1)),
+            time: self.time.unwrap_or(Time::now()),
```
```suggestion
            time: self.time.unwrap_or_else(|| Time::now()),
```

romac commented on tendermint-testgen/src/header.rs:
```diff
@@ -84,15 +84,15 @@ impl Generator<block::Header> for Header {
         };
         let  chain_id = match chain::Id::from_str(
-            choose_or(self.chain_id.clone(), "test-chain".to_string()).as_str()) {
+            self.chain_id.clone().unwrap_or("test-chain".to_string()).as_str()) {
```
```suggestion
            self.chain_id.clone().unwrap_or_else(|| "test-chain".to_string()).as_str()) {
```

andrey-kuprianov replied:
yes, right:) done

romac commented on tendermint-testgen/bin/tendermint-testgen.rs:
```diff
@@ -0,0 +1,130 @@
+    producer.encode()
+}
+
+fn run_command<Opts: Generator<T> + Options, T: serde::Serialize>(cli: Opts, read_stdin: bool) {
```
Makes the function definition a bit more readable imho.

```suggestion
fn run_command<Opts, T>(cli: Opts, read_stdin: bool)
where
    Opts: Generator<T> + Options,
    T: serde::Serialize,
{
```

romac commented on tendermint-testgen/src/vote.rs:
```diff
@@ -0,0 +1,140 @@
+        }
+    }
+    set_option!(index, u64);
+    set_option!(header, &Header, Some(header.clone()));
```
Usually preferable to leave it up to the callee to clone the data passed to a function. This can avoid unnecessary clones in cases where the callee already has ownership of the data passed to the function.

```suggestion
    set_option!(header, Header);
```

romac commented on tendermint-testgen/src/vote.rs:
```diff
@@ -0,0 +1,140 @@
+}
+
+impl Vote {
+    pub fn new(validator: &Validator, header: &Header) -> Self {
+        Vote {
+            validator: Some(validator.clone()),
+            index: None,
+            header: Some(header.clone()),
```
Usually preferable to leave it up to the callee to clone the data passed to a function. This can avoid unnecessary clones in cases where the callee already has ownership of the data passed to the function.

```suggestion
    pub fn new(validator: Validator, header: Header) -> Self {
        Vote {
            validator: Some(validator),
            index: None,
            header: Some(header),
```

romac commented on tendermint-testgen/src/commit.rs:
```diff
@@ -0,0 +1,159 @@
+}
+
+impl Commit {
+    pub fn new(header: &Header) -> Self {
+        Commit {
+            header: Some(header.clone()),
```
Usually preferable to leave it up to the callee to clone the data passed to a function. This can avoid unnecessary clones in cases where the callee already has ownership of the data passed to the function.

```suggestion
    pub fn new(header: Header) -> Self {
        Commit {
            header: Some(header),
```

andrey-kuprianov replied:
thanks:) yes; ownership and borrowing is one of the rusty things I still struggle with...

romac commented:
> Also, this needs a run of cargo fmt.

Not anymore, didn't see you pushed a commit while I was reviewing :)

romac commented:
Just want to emphasize how good the CLI interface is by the way:

```
    tendermint-testgen validator --id a --voting-power 3
     echo -n '{"id": "a", "voting_power": 3}' | tendermint-testgen --stdin validator
     echo -n a | tendermint-testgen --stdin validator --voting-power 3
     echo -n '{"id": "a"}' | tendermint-testgen --stdin validator --voting-power 3
     echo -n '{"id": "a", "voting_power": 100}' | tendermint-testgen --stdin validator --voting-power 3
```

😍

romac commented on tendermint-testgen/src/helpers.rs:
```diff
@@ -0,0 +1,64 @@
+pub fn parse_stdin_as<T: DeserializeOwned>() -> Result<T, SimpleError> {
+    let mut buffer = String::new();
+    match io::stdin().read_to_string(&mut buffer) {
+        Err(_) => Err(SimpleError::new("")),
```
It would be nice to keep the underlying error around, even if just as a string, to ease debugging a bit perhaps.

romac replied:
Also `read_stdin` and `parse_stdin_as` could perhaps be rewritten in the same style, ie. with `try_with!` or with a `match`.

andrey-kuprianov replied:
`parse_stdin_as` was not used anymore actually; so I removed it; and `read_stdin` I've rewritten using `match`

romac commented on tendermint-testgen/src/helpers.rs:
```diff
@@ -0,0 +1,64 @@
+
+pub fn read_stdin() -> Result<String, SimpleError> {
+    let mut buffer = String::new();
+    try_with!(io::stdin().read_to_string(&mut buffer), "");
```
It would be nice to keep the underlying error around, even if just as a string, to ease debugging a bit perhaps.

andrey-kuprianov replied:
yes, sure, you are right

romac commented on tendermint-testgen/src/header.rs:
```diff
@@ -0,0 +1,171 @@
+            data_hash: None,
+            validators_hash: valset.hash(),
+            next_validators_hash: next_valset.hash(),
+            consensus_hash: valset.hash(), // TODO: currently not clear how to produce a valid hash
```
@andrey-kuprianov Could you expand on this a bit?

andrey-kuprianov replied:
well, it's that from my current limited understanding of the code base I don't see an obvious way on how to compute the consensus hash... If you know the way, please share it with me

liamsi replied:
The consensus hash is the hash of the consensus params: https://github.com/informalsystems/tendermint-rs/blob/master/tendermint/src/consensus/params.rs

More precisely only Block.MaxBytes and Block.MaxGas are included in the message that gets hashed:
https://github.com/tendermint/tendermint/blob/940e68292cd2e710525c35df08d7e007a7c04f9d/types/params.go#L147-L166

(tmhash is just sha256)

andrey-kuprianov replied:
@liamsi thanks for the clarification! Still, seeing how it's implemented in Go, and with my current limited knowledge of Rust and how hashing is done in Tendermint, I don't feel comfortable enough to reproduce it, and be confident that what I write is correct. Also, in the current version of the light client this is not checked at all (should it, by the way?), as far as I can say.

If someone else volunteers to write this hashing part for consensus params, I am both hands for it! :) Otherwise, I would suggest to leave it for the future update.

ebuchman replied:
This is the hash of the consensus params which light clients currently don't pay attention to but which can change (though very infrequently) in the life of a blockchain. Best to open a follow up issue for things like this

andrey-kuprianov replied:
That's true; created a follow up issue #480.

romac commented on tendermint-testgen/src/vote.rs:
```diff
@@ -0,0 +1,165 @@
+    #[options(help = "header to sign (default: commit header)")]
+    pub header: Option<Header>,
+    #[options(help = "vote type; 'precommit' if set, otherwise 'prevote' (default)")]
+    pub precommit: Option<()>,
```
Since `gumdrop` supports boolean options, could you explain the choice of `Option<()>` instead?

andrey-kuprianov replied:
The alternatives are:
* using plain `bool`: doesn't fit well with other fields being options, so contradicts the overall architecture.
* using `Option<bool>` doesn't work either, because then `gumdrop` expects the input of the form `--precommit true` or `--precommit false`.

Using `Option<()>` on the other hand, allows in the CLI just to write `--precommit`.

andrey-kuprianov commented:
> Just want to emphasize how good the CLI interface is by the way:
> 

@romac  cool, thanks; glad you like it :) What I particularly like about `gumdrop` is that it's so transparent that I can reuse the same structs for the Rust API and for the CLI...

andrey-kuprianov commented:
@shonfeder @romac @brapse I think I've addressed all the issues raised in your reviews; thank you, that was very helpful, and allowed to improve the code a lot:)

Shivani912 commented on tendermint-testgen/src/commit.rs:
```diff
@@ -0,0 +1,190 @@
+
+    /// Generate commit votes from all validators in the header.
+    /// This function will panic if the header is not present
+    pub fn generate_default_votes(mut self) -> Self {
```
Shouldn't this be a part of `Commit::new` ? Or is there a reason we're not adding values for votes and round there?

andrey-kuprianov replied:
I don't think so. There are two scenarios we want to cover, probably:
* all (or most) votes are present, and either there are no or very little deviation in the votes from the auto-generated default votes.  In that cause the test writer will go over the route `Commit::new()` -> `generate_default_votes()` -> `modify the votes`
* the desired votes structure differs substantially from the auto-generated one. in that case the user would go over `Commit::new()` -> `add votes manually`.

In general, as a design principle of this library, we want to leave as many parameters optional as possible, but to allow the user to set them explicitely when needed. In the cause of `round`, there is a reasonable default value, `1`, which should probably suffice in most cases, that's why we leave it optional.

Shivani912 replied:
Aha, you're right. Thanks for explaining @andrey-kuprianov !

andrey-kuprianov replied:
@Shivani912 actually, you were right:) See 44c3f92.

It is a bit difficult to achieve the right balance here, because `round` and `votes` parameters of a `Commit` are not independent. Would love to hear your opinion on the new variant.

andrey-kuprianov commented on tendermint-testgen/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+edition = "2018"
+
+[dependencies]
+tendermint = { path = "../tendermint" }
```
yes, you are right :+1:

romac commented on Cargo.toml:
```diff
@@ -5,4 +5,5 @@ members = [
     "light-node",
     "rpc",
     "tendermint",
+    "tendermint-testgen"
```
How about naming the folder just `testgen` for consistency with the other crates in the workspace? We would still keep `tendermint-testgen` as the crate name in `testgen/Cargo.toml`.

andrey-kuprianov replied:
yes, nice idea, applied:)

andrey-kuprianov commented:
@brapse @romac I think I've addressed all the issues raised. For some reason Github says there are still pending changes suggested by @romac, but I don't see anything left, and don't know how to resolve this... 

If everything is OK from your side, could you please merge the PR?

andrey-kuprianov commented:
Btw, the failing `cargo fmt` has nothing to do with this PR...
